_gsocblogs/2022/blog_Acts_TirthankarMazumder.md

```diff
@@ -0,0 +1,130 @@
+---
+project: Acts 
+title: Acts — Vectorized Linear Algebra Implementation
+author: Tirthankar Mazumder
+avatar: https://avatars.githubusercontent.com/u/63574588?s=400&v=4
+date: 3.10.2022
+year: 2022
+layout: blog_post
+logo: ACTSlogo.gif
+intro: |
+  This blog is about my experiences and what I learned during my GSoC project, which involved adding a new math backend to algebra-plugins, to make Acts faster.
+---
+
+Hi everyone! Since starting this year, blogs are mandatory for all CERN-HSF contributors, I decided to take this time to make my own blog. I have written about my journey extensively there. Do [give it a visit](https://wermos.github.io/blog/)!
+
+I have broken down the midterm project report into multiple blog posts:
+
+The [first blog post](https://wermos.github.io/blog/gsoc/gsoc-first-blog-post/) discusses what GSoC is in general, for those who might just stumble upon my blog by happenstance.
+
+The [second blog post](https://wermos.github.io/blog/gsoc/gsoc-the-details-of-my-project/) delves into the technical details about what my project actually is, what repository I would be working with, mentions our weekly meeting schedule, and describes my three mentors.
+
+The [third blog post](https://wermos.github.io/blog/gsoc/gsoc-the-work-so-far/) finishes up the trilogy by exhaustively discussing the work up till now, including benchmarking data and conclusions, and the current direction of the work.
+
+That being said, I will also share a short report of the work I am doing, and the link to the things I have done so far below:
+
+## My progress
+
+### Fast 5×5
+#### Benchmarking Suite for `algebra-plugins`
+One of the main things we need, to efficiently figure out what to improve first, is a comprehensive benchmarking suite for the existing math backends. To that end, I added Google Benchmark to the [`algebra-plugins` repository](https://github.com/acts-project/algebra-plugins), in [PR #69](https://github.com/acts-project/algebra-plugins/pull/69). Writing those benchmarks is still a work-in-progress.
+
+#### Investigating Matrix Math Libraries
+Then, my mentors and I spent a lot of time testing out 3 possible math backend alternatives (the custom-made `fast5x5`, [`blaze`](https://bitbucket.org/blaze-lib/blaze/src/master/), and [`Fastor`](https://github.com/romeric/Fastor) in a repository of mine called [Fast 5×5](https://github.com/wermos/Fast5x5) and comparing their performance to that of [`Eigen`](https://eigen.tuxfamily.org/index.php?title=Main_Page). This repository was something I forked from some [prior work done on this front](https://gitlab.in2p3.fr/CodeursIntensifs/Fast5x5).
+
+#### How Matrix Math Libraries Work
+All matrix math libraries use something called [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) (short for Single Instruction, Multiple Data) to achieve a speed-up compared to a naïve implementation.
```

Comment:
All matrix math libraries use this? How do you know this?

Reply:
Well, assuming they aren't GPU or heterogeneous computing-based libraries, the alternative would be to use normal scalar instructions. This would make those libraries 4x slower, on average, than something like Fastor or blaze, on average, and hence not worth using because of their subpar performance.